### PR TITLE
[core] AddedMap::{set,get}LatLngBounds

### PR DESCRIPTION
**Pull request by brunoabinader**
Implements a setter and getter for coordinate bounds in `Map` API - which can be used e.g. to limit the boundaries in which the user can set geographical coordinates.

Related: #3602.

### CODE REVIEW

File: include/mbgl/util/geo.hpp [hunk 1]
```diff
@@ -103,14 +107,18 @@ class LatLngBounds {
...
+        LatLngBounds bounds = LatLngBounds::world();
         std::swap(bounds.sw, bounds.ne);
         return bounds;
     }
 
     // Constructs a LatLngBounds object with the tile's exact boundaries.
     LatLngBounds(const CanonicalTileID&);
 
+    explicit operator bool() const {
+        return (sw.latitude <= ne.latitude) && (sw.longitude <= ne.longitude);
```
Comment (1ec5) [inline include/mbgl/util/geo.hpp]: This would be consistent with the GeoJSON specification, but inconsistent with GeoJSONVT’s behavior of allowing a single feature or annotation to span the antimeridian (#6088), correct? If one were to [calculate the bounding box of an annotation](https://github.com/mapbox/mapbox-gl-native/blob/62ea1f21858c69f6921c775ba7a3de201f0514d8/platform/darwin/src/MGLMultiPoint.mm#L164-L167), would this operator end up evaluating to `false`? Would it still be possible to [fit the bounds to the annotation](https://github.com/mapbox/mapbox-gl-native/blob/6c92199a16a9fc6648244547a183bc80da7a56d7/platform/ios/src/MGLMapView.mm#L3960-L3972)?

**Comment by brunoabinader** [comment]
>  I have some concerns about the antimeridian edge case (the gift that keeps on giving)

Do you have one in particular? From my tests, this shouldn't be an issue as every coordinate passed to the map gets unwrapped e.g. [here](https://github.com/mapbox/mapbox-gl-native/blob/master/src/mbgl/map/transform.cpp#L91).

**Comment by brunoabinader** [comment]
> #3602 (comment) called for a setMinTilt() and setMaxTilt(), which would imply a setMinPitch() and setMaxPitch() in mbgl::Map. Would that fall within the scope of this PR?

Yep, we can certainly add it too 😃

File: include/mbgl/util/geo.hpp (hunk 1 above)
Comment (brunoabinader) [inline include/mbgl/util/geo.hpp]: My understanding of [LatLngBounds::extend](https://github.com/mapbox/mapbox-gl-native/blob/master/include/mbgl/util/geo.hpp#L129) is that it preserves the assertion above. Also, this should be fine as long as the coordinates used are unwrapped.

Nonetheless, we can add a unit test to verify 👍

**Comment by 1ec5** [comment]
>> I have some concerns about the antimeridian edge case (the gift that keeps on giving)

> Do you have one in particular? From my tests, this shouldn't be an issue as every coordinate passed to the map gets unwrapped e.g. here.

I was referring to https://github.com/mapbox/mapbox-gl-native/pull/8583#discussion_r108976652. But as long as you’ve considered this possibility, 👍.

File: include/mbgl/util/geo.hpp [hunk 2]
```diff
@@ -23,17 +23,19 @@ class LatLng {
...
 
     enum WrapMode : bool { Unwrapped, Wrapped };
 
-    LatLng(null) : latitude(std::numeric_limits<double>::quiet_NaN()), longitude(latitude) {}
+    constexpr LatLng() = default;
 
-    LatLng(double lat = 0, double lon = 0, WrapMode mode = Unwrapped)
+    constexpr LatLng(null) : latitude(std::numeric_limits<double>::quiet_NaN()), longitude(latitude) {}
+
+    constexpr LatLng(double lat, double lon, WrapMode mode = Unwrapped)
```
Comment (brunoabinader) [inline include/mbgl/util/geo.hpp]: By making `LatLngBounds::world()` constexpr [here](https://github.com/mapbox/mapbox-gl-native/pull/8583/commits/6fce4a65b242fc2d185c787104493b3c0770b511#diff-cf54d0ce1dec44ab05dcd5daf776c7f8R92), `LatLngBounds` needed a constexpr ctor, which required its members (`LatLng` for southwest and northeast bounds) to also have a constexpr ctor. We use`LatLngBounds::world()` to initialize the default `TransformState` bounds [here](https://github.com/mapbox/mapbox-gl-native/pull/8583/commits/4b8849b596fedaaeca4c7437eba9fbb45aa6f776#diff-14cb746fd86755f6100a64ac3c8ee6a9R85).

File: src/mbgl/map/transform.cpp [hunk 3]
```diff
@@ -438,12 +438,14 @@ void Transform::setScale(double scale, optional<EdgeInsets> padding, const Anima
...
-void Transform::setMinZoom(const double minZoom) {
+#pragma mark - Bounds
+
+void Transform::setMinZoom(double minZoom) {
     if (std::isnan(minZoom)) return;
     state.setMinZoom(minZoom);
 }
 
-void Transform::setMaxZoom(const double maxZoom) {
+void Transform::setMaxZoom(double maxZoom) {
```
Comment (brunoabinader) [inline src/mbgl/map/transform.cpp]: As you mentioned, there is no point in adding this to headers. Also, _const-pass-by-value_ is not a common behavior in our codebase.

For me, it is clear that a non-const _pass-by-value_ is a different object from the one it has been copied from i.e. declaring it `const` does not guarantee that the original object remains immutable.

File: src/mbgl/map/transform_state.cpp [hunk 4]
```diff
@@ -321,16 +330,18 @@ void TransformState::moveLatLng(const LatLng& latLng, const ScreenCoordinate& an
...
     Bc = newWorldSize / util::DEGREES_MAX;
     Cc = newWorldSize / util::M2PI;
 
     const double m = 1 - 1e-15;
-    const double f = util::clamp(std::sin(util::DEG2RAD * latLng.latitude), -m, m);
+    const double f = util::clamp(std::sin(util::DEG2RAD * constrained.latitude), -m, m);
 
     ScreenCoordinate point = {
-        -latLng.longitude * Bc,
+        -constrained.longitude * Bc,
```
Comment (brunoabinader) [inline src/mbgl/map/transform_state.cpp]: That's where we constrain the received `LatLng` to the specified bounds. I preferred this to be done when inserting because I suppose we set `LatLng` values much less than we request them from `TransformState`.

**Comment by tobrun** [comment]
Been testing out and adding binding integration in #8622, 
thank you for picking this up @brunoabinader.

**Comment by brunoabinader** [comment]
Removed the `constexpr`-related changes and added `{set,get}{Min,Max}Pitch`.

File: src/mbgl/map/transform.cpp (hunk 3 above)
Comment (kkaefer) [inline src/mbgl/map/transform.cpp]: Sorry, what I meant to say is that I'm doing this as a convention to avoid accidental changes to the parameter that might get read later on in the function.

File: include/mbgl/util/geo.hpp [hunk 5]
```diff
@@ -126,6 +130,16 @@ class LatLngBounds {
                       (sw.longitude + ne.longitude) / 2);
     }
 
+    LatLng constrain(const LatLng& p) const {
+        if (contains(p)) {
+            return p;
+        }
+        return LatLng {
+            p.latitude < sw.latitude ? sw.latitude : p.latitude > ne.latitude ? ne.latitude : p.latitude,
+            p.longitude < sw.longitude ? sw.longitude : p.longitude > ne.longitude ? ne.longitude : p.longitude,
```
Comment (kkaefer) [inline include/mbgl/util/geo.hpp]: Parentheses to clarify operator precedence would make the code mode readable

File: src/mbgl/map/transform.cpp (hunk 3 above)
Comment (brunoabinader) [inline src/mbgl/map/transform.cpp]: Ack - let's add them back on the definition file 👍

**Comment by brunoabinader** [comment]
@kkaefer can you please re-review?
